File: Python/arrays_and_hashing/concatenation_of_array.py
# https://leetcode.com/problems/concatenation-of-array/

# Time Complexity: O(n)
# Space Complexity: O(n)

from typing import List

class Solution:
    def getConcatenation(self, nums: List[int]) -> List[int]:
        result = []
        numsLength = len(nums)
        for i in range(2 * numsLength):
            result.append(nums[i % numsLength])
        return result

    # Alternative
    def getConcatenation2(self, nums: List[int]) -> List[int]:
        return nums + nums

    # Alternative
    def getConcatenation3(self, nums: List[int]) -> List[int]:
        res=[0]*(len(nums)*2)
        for i in range(len(nums)):
            res[i]=res[i+len(nums)]=nums[i]
        return res

File: JavaScript/arrays_and_hashing/concatenation_of_array.js
// https://leetcode.com/problems/concatenation-of-array/

// Time Complexity: O(n)
// Space Complexity: O(n)

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var getConcatenation = function (nums) {
  let res = [];
  const length = nums.length;
  for (let i = 0; i < 2 * length; i++) {
    res.push(nums[i % length]);
  }
  return res;
};

// Alternative
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var getConcatenation = function (nums) {
  const secondNums = nums;
  return nums.concat(secondNums);
};

// Alternative
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var getConcatenation = function (nums) {
  return [...nums, ...nums];
};
